{"metadata": {"kernelspec": {"display_name": "Python 3 (ipykernel)", "language": "python", "name": "python3"}, "language_info": {"codemirror_mode": {"name": "ipython", "version": 3}, "file_extension": ".py", "mimetype": "text/x-python", "name": "python", "nbconvert_exporter": "python", "pygments_lexer": "ipython3", "version": "3.7"}}, "nbformat": 4, "nbformat_minor": 2, "cells": [{"cell_type": "markdown", "metadata": {"slideshow": {"slide_type": "slide"}}, "source": ["## Visualising cell types (single cell RNA-Seq)\n", "\n", "In this practical exercise we will be exploring a biological dataset. We will be loading the single-cell RNA-seq dataset from the paper, Macosko et al, \"Highly Parallel Genome-wide Expression Profiling of Individual Cells Using Nanoliter Droplets,\" Cell (2015).\n", "\n", "It contains genetic information from 44808 individual cells, and the labels correspond to their cell types. Instead of using genes as features, the data has already been pre-processed using PCA. The 50 principal components are stored, together with 12 unique labels or cell types. \n", "\n", "We will be using t-SNE to visualise the samples and identify the clusters only based on their gene expression. The clusters should ideally correspond to the cell types.\n", "\n", "Note that for the purpose of this practical, we have randomly sampled 10,000 rows from the original dataset - this ensures we can run tSNE with different sets of parameters whilst not waiting too long. If you were to try running tSNE on the original dataset, it would usually take 20min+ to train."]}, {"cell_type": "markdown", "metadata": {"slideshow": {"slide_type": "slide"}}, "source": ["### Loading libraries and data\n", "\n", "As usual, let's start by loading in some essential libraries `pandas`, `numpy`, and `matplotlib`.\n", "\n", "To plot, in this practical we will be using `seaborn`. It is a Python data visualisation library based on `matplotlib`, which provides a high-level interface for drawing attractive and informative statistical graphics."]}, {"cell_type": "code", "execution_count": null, "metadata": {"slideshow": {"slide_type": "slide"}}, "outputs": [], "source": ["import pandas as pd\n", "import numpy as np\n", "\n", "import matplotlib.pyplot as plt\n", "%matplotlib inline\n", "\n", "from sklearn.manifold import TSNE\n", "import seaborn as sns"]}, {"cell_type": "markdown", "metadata": {"slideshow": {"slide_type": "slide"}}, "source": ["### Loading the data\n", "\n", "Our data is available as pandas dataframe in pickle format. We will load them and convert to a numpy array\n", "* Load the main dataset `data/x.pkl` into a dataframe using `pd.read_pickle`. Call `.values` on the dataframe to convert it to a numpy array and save the result in a variable called `x` \n", "* Load our labels, `data/y.pkl` into a variable (numpy array) called `y` by doing the same as above"]}, {"cell_type": "code", "execution_count": null, "metadata": {"slideshow": {"slide_type": "slide"}}, "outputs": [], "source": ["# Your code here...\n"]}, {"cell_type": "markdown", "metadata": {"slideshow": {"slide_type": "slide"}}, "source": ["### Loading the data\n", "\n", "* Inspect the data, print the number of samples and features for instance (using `.shape`)\n", "* Print the cell types or labels. You can use the method `.unique()` to identify and return a list with all unique values in an array."]}, {"cell_type": "code", "execution_count": null, "metadata": {"slideshow": {"slide_type": "slide"}}, "outputs": [], "source": ["# Your code here...\n"]}, {"cell_type": "markdown", "metadata": {"slideshow": {"slide_type": "slide"}}, "source": ["### Plot using PCA\n", "\n", "The original data was already reduced using PCA, in order to reduce the number of features. This is a common practice among biologists, as gene expression arrays can have very high dimensions - on the order of hundreds of thousands of genes. Where most genes have very low or practically zero expression. \n", "\n", "Therefore we can use the already computed PCA components to have a first look on the data. "]}, {"cell_type": "markdown", "metadata": {"slideshow": {"slide_type": "slide"}}, "source": ["### Plot using PCA\n", "\n", "* We will use `sns.scatterplot` to plot our data\n", "\n", "`seaborn` operates over `pandas` DataFrames. Define a `pandas` DataFrame with the data and use it to plot the 2 principal components (first two features of the original data)"]}, {"cell_type": "code", "execution_count": null, "metadata": {"slideshow": {"slide_type": "slide"}}, "outputs": [], "source": ["# Your code here...\n"]}, {"cell_type": "markdown", "metadata": {"slideshow": {"slide_type": "slide"}}, "source": ["### Plot using PCA\n", "\n", "* We will use `sns.scatterplot` to plot our data\n", "\n", "Now use `seaborn` to plot the DataFrame and use labels as colors. Use the pre-defined color palette:\n", "\n", "```Python\n", "    MACOSKO_COLORS = {\n", "        \"Amacrine cells\": \"#A5C93D\",\n", "        \"Astrocytes\": \"#8B006B\",\n", "        \"Bipolar cells\": \"#2000D7\",\n", "        \"Cones\": \"#538CBA\",\n", "        \"Fibroblasts\": \"#8B006B\",\n", "        \"Horizontal cells\": \"#B33B19\",\n", "        \"Microglia\": \"#8B006B\",\n", "        \"Muller glia\": \"#8B006B\",\n", "        \"Pericytes\": \"#8B006B\",\n", "        \"Retinal ganglion cells\": \"#C38A1F\",\n", "        \"Rods\": \"#538CBA\",\n", "        \"Vascular endothelium\": \"#8B006B\",\n", "    }\n", "````\n", "\n", "\n"]}, {"cell_type": "code", "execution_count": null, "metadata": {"slideshow": {"slide_type": "slide"}}, "outputs": [], "source": ["# Your code here...\n"]}, {"cell_type": "markdown", "metadata": {"slideshow": {"slide_type": "slide"}}, "source": ["We can see that even though the first 2 principal components account for most of the data variability, they are still not enough to provide a visualisation with separable clusters or cell types.\n"]}, {"cell_type": "markdown", "metadata": {"slideshow": {"slide_type": "slide"}}, "source": ["### Run t-SNE\n", "\n", "Instantiate the `TSNE` class, fit and transform the data. \n", "\n", "Set `verbose = True` to print some of the operation details while training. \n", "Set `random_state = 0` to initialize the internal random number generator equally for all models."]}, {"cell_type": "code", "execution_count": null, "metadata": {"slideshow": {"slide_type": "slide"}}, "outputs": [], "source": ["# Your code here...\n"]}, {"cell_type": "code", "execution_count": null, "metadata": {"slideshow": {"slide_type": "slide"}}, "outputs": [], "source": ["%time tsne_embedding = tsne.fit_transform(x)"]}, {"cell_type": "markdown", "metadata": {"slideshow": {"slide_type": "slide"}}, "source": ["### Plot t-SNE\n", "\n", "Similarly to what we did before, define and execute a function that uses the t-SNE embedding to plot the data using `seaborn`."]}, {"cell_type": "code", "execution_count": null, "metadata": {"slideshow": {"slide_type": "slide"}}, "outputs": [], "source": ["# Your code here...\n"]}, {"cell_type": "markdown", "metadata": {"slideshow": {"slide_type": "slide"}}, "source": ["### Explore different numbers of iterations\n", "\n", "Let's see what the embeddings look like at different stages of training. This is controlled by the number of iterations that we allow the algorithm to run for. \n", "\n", "* We can modify the number of algorithm iterations by setting `n_iter` \n", "\n", "When the number of iterations is not defined, the default is to run until convergence. \n", "\n", "More information can be found on the official `sklearn` documentation at https://scikit-learn.org/stable/modules/generated/sklearn.manifold.TSNE.html "]}, {"cell_type": "code", "execution_count": null, "metadata": {"slideshow": {"slide_type": "slide"}}, "outputs": [], "source": ["# Your code here...\n"]}, {"cell_type": "markdown", "metadata": {"slideshow": {"slide_type": "slide"}}, "source": ["### Explore different perplexities\n", "\n", "The `perplexity` hyper-parameter is directly related to the number of neighbours considered for each sample during training. \n", "\n", "Therefore it has an effect over the local vs global structure of the t-SNE embedding.\n", "\n", "Try different values of `perplexity` and observe their effect on the t-SNE embedding. Modify the `perplexity` value inside TSNE\n", "\n", "The default perplexity value is `p=30`."]}, {"cell_type": "code", "execution_count": null, "metadata": {"slideshow": {"slide_type": "slide"}}, "outputs": [], "source": ["# Your code here...\n"]}]}